### ISSUE TITLE
weight decay

### ISSUE THREAD

**Issue by tycho01** [hasktorch/hasktorch#452]
I'd tried my hand at un-typed weight decay as follows:
```
import qualified Torch.Functional              as F
import qualified Torch.Tensor                  as D
import qualified Torch.Optim                   as D

decayWeights :: Float -> [D.IndependentTensor] -> D.Gradients -> D.Gradients
decayWeights v parameters (D.Gradients gradients) = D.Gradients $ zipWith (\ param gradient -> (if v == (0.0 :: Float) then id else F.add (F.mulScalar v param)) gradient) (D.toDependent <$> parameters) gradients
```

my optimization then looked kinda like this:
```
D.runStep' model optim (toDynamic lr) . decayWeights regularization params $ D.grad (toDynamic loss) params
        where params = A.flattenParameters model
```

Problem being, adding this function call into my optimization line caused a massive memory leak that would soon kill my training loop.
I've no idea what caused this, so I guess I'm curious if this was just me or if it's an actual issue with my `decayWeights` implementation...

**Comment by junjihashimoto** [comment]
https://github.com/hasktorch/hasktorch/pull/453
If you try to use this and a leak occurs, there may be  unfreeed tensors.
